File: .claude/skills/linkedin-post/scripts/publisher.py
#!/usr/bin/env python3
"""
linkedin_publisher.py — Pubblica post su LinkedIn via API v2
Uso:
  publisher.py --dry-run                                  Testa autenticazione
  publisher.py --text "testo" [--image path/img.jpg]      Post singolo
  publisher.py --text "testo" --carousel img1 img2 img3   Post carosello
"""

import argparse
import json
import os
import sys
from pathlib import Path

import requests
from dotenv import load_dotenv

load_dotenv(Path(__file__).parents[5] / ".env")

ACCESS_TOKEN = os.getenv("LINKEDIN_ACCESS_TOKEN")
PERSON_ID = os.getenv("LINKEDIN_PERSON_ID")
API_BASE = "https://api.linkedin.com/v2"


def _headers():
    return {
        "Authorization": f"Bearer {ACCESS_TOKEN}",
        "Content-Type": "application/json",
        "X-Restli-Protocol-Version": "2.0.0",
    }


def _check_credentials():
    if not ACCESS_TOKEN or not PERSON_ID:
        print("❌ Credenziali LinkedIn mancanti. Compila LINKEDIN_ACCESS_TOKEN e LINKEDIN_PERSON_ID in .env", file=sys.stderr)
        print("   Guida: .claude/setup/api-credentials.md", file=sys.stderr)
        sys.exit(1)


def dry_run():
    _check_credentials()
    r = requests.get(f"{API_BASE}/userinfo", headers=_headers())
    if r.status_code == 200:
        data = r.json()
        print(f"✓ Auth OK — dry run completato. Account: {data.get('name', 'N/A')}")
    else:
        print(f"❌ Auth fallita: {r.status_code} {r.text}", file=sys.stderr)
        sys.exit(1)


def _register_upload(upload_type="feedshare-image"):
    payload = {
        "registerUploadRequest": {
            "recipes": [f"urn:li:digitalmediaRecipe:{upload_type}"],
            "owner": f"urn:li:person:{PERSON_ID}",
            "serviceRelationships": [
                {"relationshipType": "OWNER", "identifier": "urn:li:userGeneratedContent"}
            ],
        }
    }
    r = requests.post(f"{API_BASE}/assets?action=registerUpload", headers=_headers(), json=payload)
    r.raise_for_status()
    data = r.json()
    upload_url = data["value"]["uploadMechanism"]["com.linkedin.digitalmedia.uploading.MediaUploadHttpRequest"]["uploadUrl"]
    asset = data["value"]["asset"]
    return upload_url, asset


def upload_image(image_path):
    upload_url, asset = _register_upload()
    with open(image_path, "rb") as f:
        img_data = f.read()
    r = requests.put(upload_url, data=img_data, headers={"Authorization": f"Bearer {ACCESS_TOKEN}"})
    r.raise_for_status()
    print(f"✓ Immagine caricata: {asset}")
    return asset


def publish_post(text, asset_urn=None, carousel_urns=None):
    body = {
        "author": f"urn:li:person:{PERSON_ID}",
        "lifecycleState": "PUBLISHED",
        "specificContent": {
            "com.linkedin.ugc.ShareContent": {
                "shareCommentary": {"text": text},
                "shareMediaCategory": "NONE" if not asset_urn and not carousel_urns else "IMAGE",
            }
        },
        "visibility": {"com.linkedin.ugc.MemberNetworkVisibility": "PUBLIC"},
    }

    if asset_urn:
        body["specificContent"]["com.linkedin.ugc.ShareContent"]["shareMediaCategory"] = "IMAGE"
        body["specificContent"]["com.linkedin.ugc.ShareContent"]["media"] = [
            {"status": "READY", "media": asset_urn}
        ]
    elif carousel_urns:
        body["specificContent"]["com.linkedin.ugc.ShareContent"]["shareMediaCategory"] = "IMAGE"
        body["specificContent"]["com.linkedin.ugc.ShareContent"]["media"] = [
            {"status": "READY", "media": urn} for urn in carousel_urns
        ]

    r = requests.post(f"{API_BASE}/ugcPosts", headers=_headers(), json=body)
    r.raise_for_status()
    post_id = r.headers.get("x-restli-id", r.json().get("id", ""))
    post_url = f"https://www.linkedin.com/feed/update/{post_id}/"
    print(f"✓ Post pubblicato: {post_url}")
    return post_url


def main(text, image_path=None, carousel_paths=None):
    _check_credentials()
    asset_urn = None
    carousel_urns = None

    if image_path:
        asset_urn = upload_image(image_path)
    elif carousel_paths:
        carousel_urns = [upload_image(p) for p in carousel_paths]

    return publish_post(text, asset_urn=asset_urn, carousel_urns=carousel_urns)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="LinkedIn Publisher")
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--text", type=str)
    parser.add_argument("--image", type=str)
    parser.add_argument("--carousel", nargs="+", metavar="IMG")
    args = parser.parse_args()

    if args.dry_run:
        dry_run()
    elif args.text:
        main(args.text, image_path=args.image, carousel_paths=args.carousel)
    else:
        parser.print_help()

File: .claude/skills/linkedin-post/scripts/publisher.js
#!/usr/bin/env node
/**
 * linkedin_publisher.js — Pubblica post su LinkedIn via API v2
 * Uso:
 *   node publisher.js --dry-run
 *   node publisher.js --text "testo" [--image path/img.jpg]
 *   node publisher.js --text "testo" --carousel img1 img2 img3
 */

import { readFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { config } from 'dotenv';

const __dirname = dirname(fileURLToPath(import.meta.url));
config({ path: resolve(__dirname, '../../../../.env') });

const ACCESS_TOKEN = process.env.LINKEDIN_ACCESS_TOKEN;
const PERSON_ID = process.env.LINKEDIN_PERSON_ID;
const API_BASE = 'https://api.linkedin.com/v2';

function headers() {
  return {
    'Authorization': `Bearer ${ACCESS_TOKEN}`,
    'Content-Type': 'application/json',
    'X-Restli-Protocol-Version': '2.0.0',
  };
}

function checkCredentials() {
  if (!ACCESS_TOKEN || !PERSON_ID) {
    console.error('❌ Credenziali LinkedIn mancanti. Compila LINKEDIN_ACCESS_TOKEN e LINKEDIN_PERSON_ID in .env');
    console.error('   Guida: .claude/setup/api-credentials.md');
    process.exit(1);
  }
}

async function dryRun() {
  checkCredentials();
  const r = await fetch(`${API_BASE}/userinfo`, { headers: headers() });
  if (r.ok) {
    const data = await r.json();
    console.log(`✓ Auth OK — dry run completato. Account: ${data.name ?? 'N/A'}`);
  } else {
    console.error(`❌ Auth fallita: ${r.status} ${await r.text()}`);
    process.exit(1);
  }
}

async function registerUpload() {
  const payload = {
    registerUploadRequest: {
      recipes: ['urn:li:digitalmediaRecipe:feedshare-image'],
      owner: `urn:li:person:${PERSON_ID}`,
      serviceRelationships: [{ relationshipType: 'OWNER', identifier: 'urn:li:userGeneratedContent' }],
    },
  };
  const r = await fetch(`${API_BASE}/assets?action=registerUpload`, {
    method: 'POST', headers: headers(), body: JSON.stringify(payload),
  });
  if (!r.ok) throw new Error(`registerUpload fallito: ${r.status} ${await r.text()}`);
  const data = await r.json();
  const uploadUrl = data.value.uploadMechanism['com.linkedin.digitalmedia.uploading.MediaUploadHttpRequest'].uploadUrl;
  const asset = data.value.asset;
  return { uploadUrl, asset };
}

async function uploadImage(imagePath) {
  const { uploadUrl, asset } = await registerUpload();
  const imgData = readFileSync(imagePath);
  const r = await fetch(uploadUrl, {
    method: 'PUT',
    body: imgData,
    headers: { 'Authorization': `Bearer ${ACCESS_TOKEN}` },
  });
  if (!r.ok) throw new Error(`Upload immagine fallito: ${r.status}`);
  console.log(`✓ Immagine caricata: ${asset}`);
  return asset;
}

async function publishPost(text, assetUrn = null, carouselUrns = null) {
  const media = assetUrn
    ? [{ status: 'READY', media: assetUrn }]
    : carouselUrns
    ? carouselUrns.map(urn => ({ status: 'READY', media: urn }))
    : null;

  const body = {
    author: `urn:li:person:${PERSON_ID}`,
    lifecycleState: 'PUBLISHED',
    specificContent: {
      'com.linkedin.ugc.ShareContent': {
        shareCommentary: { text },
        shareMediaCategory: media ? 'IMAGE' : 'NONE',
        ...(media && { media }),
      },
    },
    visibility: { 'com.linkedin.ugc.MemberNetworkVisibility': 'PUBLIC' },
  };

  const r = await fetch(`${API_BASE}/ugcPosts`, {
    method: 'POST', headers: headers(), body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(`Publish fallito: ${r.status} ${await r.text()}`);
  const postId = r.headers.get('x-restli-id') ?? (await r.json()).id ?? '';
  const postUrl = `https://www.linkedin.com/feed/update/${postId}/`;
  console.log(`✓ Post pubblicato: ${postUrl}`);
  return postUrl;
}

async function main(text, imagePath = null, carouselPaths = null) {
  checkCredentials();
  let assetUrn = null;
  let carouselUrns = null;
  if (imagePath) assetUrn = await uploadImage(imagePath);
  else if (carouselPaths) carouselUrns = await Promise.all(carouselPaths.map(uploadImage));
  return publishPost(text, assetUrn, carouselUrns);
}

// CLI
const args = process.argv.slice(2);
const get = (flag) => { const i = args.indexOf(flag); return i !== -1 ? args[i + 1] : null; };
const getAll = (flag) => { const i = args.indexOf(flag); if (i === -1) return null; const vals = []; let j = i + 1; while (j < args.length && !args[j].startsWith('--')) vals.push(args[j++]); return vals.length ? vals : null; };

if (args.includes('--dry-run')) {
  dryRun().catch(e => { console.error(e.message); process.exit(1); });
} else if (get('--text')) {
  main(get('--text'), get('--image'), getAll('--carousel')).catch(e => { console.error(e.message); process.exit(1); });
} else {
  console.log('Uso: node publisher.js --dry-run | --text "testo" [--image path] [--carousel img1 img2]');
}
